File: docroot/WEB-INF/src/com/marcelmika/lims/portal/domain/Properties.java
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Marcel Mika, marcelmika.com
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.marcelmika.lims.portal.domain;

import com.liferay.compat.portal.kernel.util.StringUtil;
import com.marcelmika.lims.api.environment.Environment;
import com.marcelmika.lims.api.environment.Environment.BuddyListStrategy;
import com.marcelmika.lims.api.environment.Environment.BuddyListSocialRelation;

/**
 * @author Ing. Marcel Mika
 * @link http://marcelmika.com
 * Date: 08/10/14
 * Time: 21:39
 */
public class Properties {

    private BuddyListStrategy buddyListStrategy;
    private BuddyListSocialRelation[] buddyListSocialRelations;
    private Boolean buddyListIgnoreDefaultUser;
    private Boolean buddyListIgnoreDeactivatedUser;
    private Integer buddyListMaxBuddies;
    private Integer buddyListMaxSearch;
    private Integer conversationListMaxMessages;
    private String[] buddyListSiteExcludes;
    private String[] buddyListGroupExcludes;

    /**
     * Factory method that creates an instance of properties from the environment properties
     *
     * @return Properties
     */
    public static Properties fromEnvironment() {
        // Create new instance
        Properties properties = new Properties();

        // Map properties
        properties.buddyListStrategy = Environment.getBuddyListStrategy();
        properties.buddyListSocialRelations = Environment.getBuddyListSocialRelations();
        properties.buddyListIgnoreDefaultUser = Environment.getBuddyListIgnoreDefaultUser();
        properties.buddyListIgnoreDeactivatedUser = Environment.getBuddyListIgnoreDeactivatedUser();
        properties.buddyListMaxBuddies = Environment.getBuddyListMaxBuddies();
        properties.buddyListMaxSearch = Environment.getBuddyListMaxSearch();
        properties.conversationListMaxMessages = Environment.getConversationListMaxMessages();
        properties.buddyListSiteExcludes = Environment.getBuddyListSiteExcludes();
        properties.buddyListGroupExcludes = Environment.getBuddyListGroupExcludes();

        return properties;
    }

    public BuddyListStrategy getBuddyListStrategy() {
        return buddyListStrategy;
    }

    public void setBuddyListStrategy(BuddyListStrategy buddyListStrategy) {
        this.buddyListStrategy = buddyListStrategy;
    }

    public BuddyListSocialRelation[] getBuddyListSocialRelations() {
        return buddyListSocialRelations;
    }

    public void setBuddyListSocialRelations(BuddyListSocialRelation[] buddyListSocialRelations) {
        this.buddyListSocialRelations = buddyListSocialRelations;
    }

    public Boolean getBuddyListIgnoreDefaultUser() {
        return buddyListIgnoreDefaultUser;
    }

    public void setBuddyListIgnoreDefaultUser(Boolean buddyListIgnoreDefaultUser) {
        this.buddyListIgnoreDefaultUser = buddyListIgnoreDefaultUser;
    }

    public Boolean getBuddyListIgnoreDeactivatedUser() {
        return buddyListIgnoreDeactivatedUser;
    }

    public void setBuddyListIgnoreDeactivatedUser(Boolean buddyListIgnoreDeactivatedUser) {
        this.buddyListIgnoreDeactivatedUser = buddyListIgnoreDeactivatedUser;
    }

    public Integer getBuddyListMaxBuddies() {
        return buddyListMaxBuddies;
    }

    public void setBuddyListMaxBuddies(Integer buddyListMaxBuddies) {
        this.buddyListMaxBuddies = buddyListMaxBuddies;
    }

    public Integer getBuddyListMaxSearch() {
        return buddyListMaxSearch;
    }

    public void setBuddyListMaxSearch(Integer buddyListMaxSearch) {
        this.buddyListMaxSearch = buddyListMaxSearch;
    }

    public Integer getConversationListMaxMessages() {
        return conversationListMaxMessages;
    }

    public void setConversationListMaxMessages(Integer conversationListMaxMessages) {
        this.conversationListMaxMessages = conversationListMaxMessages;
    }

    public String[] getBuddyListSiteExcludes() {
        return buddyListSiteExcludes;
    }

    public String getBuddyListSiteExcludesFlatten() {
        return StringUtil.merge(buddyListSiteExcludes, ",");
    }

    public void setBuddyListSiteExcludes(String[] buddyListSiteExcludes) {
        this.buddyListSiteExcludes = buddyListSiteExcludes;
    }

    public String[] getBuddyListGroupExcludes() {
        return buddyListGroupExcludes;
    }

    public String getBuddyListGroupExcludesFlatten() {
        return StringUtil.merge(buddyListGroupExcludes, ",");
    }

    public void setBuddyListGroupExcludes(String[] buddyListGroupExcludes) {
        this.buddyListGroupExcludes = buddyListGroupExcludes;
    }
}

File: docroot/js/src/lims/js/core/properties/Properties.js
/*
 * The MIT License (MIT)
 *
 * Copyright (c) 2014 Marcel Mika, marcelmika.com
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

/**
 * Settings
 *
 * Contains all necessary settings related to the current user session
 */
Y.namespace('LIMS.Core');

Y.LIMS.Core.Properties = Y.Base.create('properties', Y.Base, [], {

    userId: null,      // This is set in main.js, to access it use Y.LIMS.Core.Properties.userId
    companyId: null,   // This is set in main.js, to access it use Y.LIMS.Core.Properties.companyId
    pathImage: null,   // This is set in main.js, to access it use Y.LIMS.Core.Properties.pathImage
    isIE: false,       // This is set in main.js, to access it use Y.LIMS.Core.Properties.isIE
    resourceURL: null,  // This is set in main.js, to access it use Y.LIMS.Core.Properties.resourceURL

    /**
     * Values are parsed from HTML in main.js.
     * Check the properties.jspf for the list of all possible properties.
     *
     * {}
     */
    values: null,

    /**
     * Called when the object is created
     */
    initializer: function () {
        // Attach to events
        this._attachEvents();
    },

    /**
     * Returns userId of the currently connected user
     *
     * @returns {int}
     */
    getCurrentUserId: function () {
        return this.get('userId');
    },

    /**
     * Returns companyId of the currently connected user
     *
     * @return {number}
     */
    getCurrentUserCompanyId: function () {
        return this.get('companyId');
    },

    /**
     * Returns true if the current user is male
     *
     * @return {boolean}
     */
    getCurrentUserMale: function () {
        return this.get('male');
    },

    /**
     * Returns portraitId of the currently connected user
     *
     * @return {number}
     */
    getCurrentUserPortraitId: function () {
        return this.get('portraitId');
    },

    /**
     * Returns portrait image token of the currently connected user
     *
     * @return {string}
     */
    getCurrentUserPortraitImageToken: function () {
        return this.get('portraitImageToken');
    },

    /**
     * Returns portrait token of the currently connected user
     *
     * @return {string}
     */
    getCurrentUserPortraitToken: function () {
        return this.get('portraitToken');
    },

    /**
     * Returns screenName of the currently connected user
     *
     * @returns {string}
     */
    getCurrentUserScreenName: function () {
        return this.get('screenName');
    },

    /**
     * Returns fullName of the currently connected user
     *
     * @returns {string}
     */
    getCurrentUserFullName: function () {
        return this.get('fullName');
    },


    /**
     * Returns server time offset compared to client time
     *
     * @returns {timestamp}
     */
    getServerTimeOffset: function () {
        return this.get('offset');
    },

    /**
     * Returns true if the whole chat is enabled
     *
     * @returns {boolean}
     */
    isChatEnabled: function () {
        return this.get('isChatEnabled');
    },

    /**
     * Attach dom events
     *
     * @private
     */
    _attachEvents: function () {
        // Chat enabled/disabled
        Y.on('chatEnabled', this._onChatEnabled, this);
        Y.on('chatDisabled', this._onChatDisabled, this);
    },

    /**
     * Called when the chat is enabled (turned on)
     *
     * @private
     */
    _onChatEnabled: function () {
        this.set('isChatEnabled', true);
    },

    /**
     * Called when the chat is disabled (turned off)
     *
     * @private
     */
    _onChatDisabled: function () {
        this.set('isChatEnabled', false);
    }

}, {

    ATTRS: {

        /**
         * User Id of the currently logged user
         *
         * {integer}
         */
        userId: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.userId;
            }
        },

        /**
         * Company Id of the currently logged user
         *
         * {integer}
         */
        companyId: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.companyId;
            }
        },

        /**
         * Set to true if the user is male
         *
         * {boolean}
         */
        male: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.male;
            }
        },

        /**
         * Portrait id of the currently logged user
         *
         * {integer}
         */
        portraitId: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.portraitId;
            }
        },

        /**
         * Portrait image token of the currently logged user
         *
         * {string}
         */
        portraitImageToken: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.portraitImageToken;
            }
        },

        /**
         * Portrait token of the currently logged user
         *
         * {string}
         */
        portraitToken: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.portraitToken;
            }
        },

        /**
         * Screen name of the currently logged user
         *
         * {string}
         */
        screenName: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.screenName;
            }
        },


        /**
         * Full name of the currently logged user
         *
         * {string}
         */
        fullName: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.fullName;
            }
        },

        /**
         * Holds current server time
         *
         * {number}
         */
        serverTime: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.currentServerTime;
            }
        },

        /**
         * Holds delta between current server time and client time
         */
        offset: {
            valueFn: function () {
                // Get server time
                var currentChatServerTime = this.get('serverTime');
                // Count offset
                return new Date().getTime() - currentChatServerTime;
            }
        },

        /**
         * Path to images
         */
        pathImage: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.pathImage;
            }
        },

        /**
         * Set to true if chat is enabled
         *
         * {boolean}
         */
        isChatEnabled: {
            valueFn: function () {
                return Y.LIMS.Core.Properties.values.portletEnabled;
            }
        }
    }
});
